OptimizelySDK.Tests/OptimizelyTest.cs

```diff
@@ -1109,5 +1120,350 @@ public void TestGetVariationBucketingIdAttribute()
+            Assert.False(Optimizely.IsFeatureEnabled("", TestUserId, null));
+
+            LoggerMock.Verify(l => l.Log(LogLevel.ERROR, "User ID must not be empty."), Times.Exactly(2));
+            LoggerMock.Verify(l => l.Log(LogLevel.ERROR, "Feature flag key must not be empty."), Times.Exactly(2));
+        }
+
+        // Should return false and log error message when feature flag key is invalid.
+        [Test]
+        public void TestIsFeatureEnabledGivenFeatureFlagNotFound()
+        {
+            var featureKey = "this_feature_should_never_be_found_in_the_datafile_unless_the_datafile_creator_got_insane";
+            Assert.False(Optimizely.IsFeatureEnabled(featureKey, TestUserId, null));
+            
+            LoggerMock.Verify(l => l.Log(LogLevel.ERROR, $@"Feature key ""{featureKey}"" is not in datafile."));
+        }
+
+        // Should return false and log error message when arguments are null or empty.
+        [Test]
+        public void TestIsFeatureEnabledGivenFeatureFlagContainsInvalidExperiment()
+        {
+            var tempConfig = ProjectConfig.Create(TestData.Datafile, LoggerMock.Object, new NoOpErrorHandler());
+            var featureFlag = tempConfig.GetFeatureFlagFromKey("multi_variate_feature");
+
+            var optly = Helper.CreatePrivateOptimizely();
+            optly.SetFieldOrProperty("Config", tempConfig);
+
+            // Set such an experiment to the list of experiment ids, that does not belong to the feature.
+            featureFlag.ExperimentIds = new List<string> { "4209211" };
+
+            // Should return false when the experiment in feature flag does not get found in the datafile.
+            Assert.False((bool)optly.Invoke("IsFeatureEnabled", "multi_variate_feature", TestUserId, null));
+
+            LoggerMock.Verify(l => l.Log(LogLevel.ERROR, @"Experiment ID ""4209211"" is not in datafile."));
+        }
+
+        // Should return default value and log message when feature is not enabled for the user.
+        [Test]
+        public void TestIsFeatureEnabledGivenFeatureFlagIsNotEnabledForUser()
+        {
+            var featureKey = "double_single_variable_feature";
+            var featureFlag = Config.GetFeatureFlagFromKey("double_single_variable_feature");
+            
+            DecisionServiceMock.Setup(ds => ds.GetVariationForFeature(featureFlag, TestUserId, null)).Returns<Variation>(null);
+
+            var optly = Helper.CreatePrivateOptimizely();
+            optly.SetFieldOrProperty("DecisionService", DecisionServiceMock.Object);
+
+            bool result = (bool)optly.Invoke("IsFeatureEnabled", featureKey, TestUserId, null);
+            Assert.False(result);
+
+            LoggerMock.Verify(l => l.Log(LogLevel.INFO,
+                $@"Feature flag ""{featureKey}"" is not enabled for user ""{TestUserId}""."));
+        }
+
+        // Should return true when feature is enabled for the user but user does not get experimented.
```

Comment:
Should return true but not send an impression event when feature is enabled for the user but user does not get experimented.

Reply:
Done